src/SimCan.cpp

```diff
@@ -15,53 +15,106 @@ void SimCan::begin(){
     uint8_t error = _can->begin(CAN_500KBPS,MCP_8MHz);
 
     if(_serial){
-        _serial->println("CAN Init Status: " + getErrorDescription(error));
+        _serial->println("CAN Init Status: " + _getErrorDescription(error));
     }
 
-    _can_last_update = millis();
+    _last_transmit = millis();
 }
 
 void SimCan::handle(){
     // Send CAN Messages
-    if (millis() - _can_last_update >= CAN_UPDATE_MS) {
-        _can_last_update = millis();
-        uint8_t error1 = _can->sendMsgBuf(CAN0_ID, CAN_FRAME, CAN0_DATA_LENGTH, _can0_data);  
-        uint8_t error2 = _can->sendMsgBuf(CAN1_ID, CAN_FRAME, CAN1_DATA_LENGTH, _can1_data);  
-        
-        if(_serial){
-            _serial->println("CAN Msg 1 Status: " + getErrorDescription(error1));
-            _serial->println("CAN Msg 2 Status: " + getErrorDescription(error2));
-        }
+    if (millis() - _last_transmit >= CAN_TRANSMIT_INTERVAL) {
+        _last_transmit = millis();
+        _transmit();
     } 
 
-    // Print any CAN messages received (debug mode only)
+    // Listen for CAN messages
+    if (_can->checkReceive() == CAN_MSGAVAIL) {
+        _receive();
+    }
+
+}
+
+void SimCan::_transmit() {
+        for(CanMessage msg : TRANSMIT_MSGS){
+            uint8_t error = _can->sendMsgBuf(msg.id, CAN_FRAME, msg.dataLength, msg.data);
+            if(_serial){
+                _serial->print("CAN MESSAGE SENT - ID: 0x"); 
+                _serial->print(msg.id, HEX); 
+                _serial->println(" - Status: " + _getErrorDescription(error));
+            }
+        }
+
+}
+
+void SimCan::_receive(){
+    uint8_t len = 0;
+    uint8_t buf[8];
+
+    _can->readMsgBuf(&len, buf); 
+
+    uint16_t canId = _can->getCanId();
+
+    if(canId == BMS_REQUEST_ID){
+        _processBmsRequest(len, buf);
+    }
+
     if(_serial){
-        unsigned char len = 0;
-        unsigned char buf[8];
+        _serial->println("-----------------------------");
+        _serial->print("CAN MESSAGE RECEIVED - ID: 0x");
+        _serial->println(canId, HEX);
 
-        if (_can->checkReceive() == CAN_MSGAVAIL) {
-            _can->readMsgBuf(&len, buf); 
+        for (int i = 0; i < len; i++) { // print the data
+            _serial->print("0x");
+            _serial->print(buf[i], HEX);
+            _serial->print("\t");
+        }
+        _serial->println();
+    }
+}
 
-            unsigned long canId = _can->getCanId();
+void SimCan::_processBmsRequest(uint8_t len, uint8_t buf[]){
 
-            _serial->println("-----------------------------");
-            _serial->print("Get data from ID: 0x");
-            _serial->println(canId, HEX);
+    if(_serial) _serial->println("BMS CAN REQUEST RECEIVED!");
+    
+    // Check to make sure request length is correct
+    if (len != 8 && _serial){
```

Comment:
(I haven't looked at the documentation for this)

Will all messages have 8 bytes? It might be worth making this a macro.

Reply:
Yes, all the requests have 8 bytes of data. Fixed!